README.md
# RailsCallbackLog

Do you have a rails app with a lot of callbacks? Are they kind of a mystery? Maybe logging them would help.

```
Started GET "/" for 127.0.0.1 at 2016-07-26 13:25:32 -0400
Processing by HomeController#index as HTML
Callback: verify_authenticity_token
Callback: activate_authlogic
Callback: require_client_subdomain
  Client Load (0.4ms)  SELECT  `clients`.* ...
Callback: check_hostname
Callback: update_last_request_at
...
```

## Installation

```ruby
# Gemfile
gem "rails-callback_log", group: [:development, :test]
```

Do not use this gem in production because it adds significant overhead.

## Filtering Output

Rails has a lot of its own callbacks that you probably don't care about. If you
don't want to log them, enable filtering.

```bash
# Enable filtering
export RAILS_CALLBACK_LOG_FILTER="make it so"

# Disable filtering
unset RAILS_CALLBACK_LOG_FILTER
```

Filtering incurs a serious performance penalty, so it is off by default.

## Team Opt-in

When working in a team and logging is needed only for select members, use this
environment-based opt-in pattern:

```ruby
# Gemfile
gem "rails-callback_log",
  group: [:development, :test],
  require: (ENV["LOG_RAILS_CALLBACKS"] == "true")
```

## See Also

- http://stackoverflow.com/q/13089936/567762

## License

The gem is available as open source under the terms of the [MIT
License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog

This gem conforms to [semver 2.0.0][1] and follows the recommendations of
[keepachangelog.com][2].

### Unreleased

Breaking Changes:

- None

Added:

- None

Fixed:

- None

### 0.3.1 (2019-12-11)

Breaking Changes:

- None

Added:

- Relax dependency constraint on activesupport, removing upper bound

Fixed:

- None

### 0.3.0 (2019-01-29)

Breaking Changes:

- Drop `RailsCallbackLog::VERSION`, use `RailsCallbackLog.gem_version`

Added:

- Support rails 5.2

Fixed:

- [#5](https://github.com/jaredbeck/rails-callback_log/pull/5)
  Huge improvements to output in rails > 5.1

### 0.2.2 (2017-05-08)

Breaking Changes:

- None

Added:

- None

Fixed:

- [#3](https://github.com/jaredbeck/rails-callback_log/pull/3)
  Fallback to ::Logger if Rails::Logger not loaded

### 0.2.1 (2017-05-02)

Breaking Changes:

- None

Added:

- None

Fixed:

- Fix a performance issue introduced in 0.2.0

### 0.2.0 (2017-05-01)

Breaking Changes:

- None

Added:

- Support for rails 5.1

### 0.1.0 (2016-07-25)

Breaking changes:

- Drop support for ruby 1.9.3

Added:

- Support for rails 5.0

### 0.0.3 (2016-06-24)

Initial release, support for rails 4.2 only.

[1]: http://semver.org/
[2]: http://keepachangelog.com/


spec/output_spec.rb
require "spec_helper"

class CallbackWrapper
  def after(record)
    Rails.logger.info("- in CallbackWrapper after method")
  end
end

class CallbackTester
  include ActiveSupport::Callbacks
  define_callbacks :save

  def save
    run_callbacks :save do
      Rails.logger.info("- save")
    end
  end

  # Set a proc based callback
  set_callback :save, :before do
    Rails.logger.info("- in before save callback")
  end

  # Set an object based callback, will call `after` on CallbackWrapper
  set_callback :save, :after, CallbackWrapper.new

  # Set a symbol based method reference callback
  set_callback :save, :around, :around_callback
  def around_callback
    Rails.logger.info("- in around callback before")
    yield
    Rails.logger.info("- in around callback after")
  end
end

describe RailsCallbackLog do
  # Filter the logs down to only the lines that start with Callback. Makes the
  # test output easier to read.
  subject(:filtered_log) do
    output.string.split("\n").select { |line| line =~ /^Callback/ }.join("\n")
  end

  # Create a logger that logs to a StringIO that we can test against
  let(:output) { StringIO.new }
  let(:logger) do
    logger = Logger.new(output)
    logger.formatter = proc { |severity, datetime, progname, msg|
      "#{msg}\n"
    }
    logger
  end

  before do
    # Set the logger to our custom StringIO logger
    Rails.logger = logger
  end

  describe ".log" do
    it "logs all of the callbacks" do
      CallbackTester.new.save

      expect(filtered_log).to include "Callback: #<Proc:"
      expect(filtered_log).to include "Callback: :around_callback"
      expect(filtered_log).to include "Callback: #<CallbackWrapper:"
    end
  end
end


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)

# When we load `rails-callback_log.rb`, it will want to know what version of rails we
# is loaded. We don't want to load all of rails just to run our tests, so we simply
# define `::Rails.gem_version`.
module Rails
  class << self
    attr_accessor :logger
  end

  def self.gem_version
    ::ActiveSupport.gem_version
  end
end

require "rails-callback_log"


spec/rails-callback_log_spec.rb
require "spec_helper"

module ActiveSupport
  module Callbacks
    if ::ActiveSupport.gem_version >= ::Gem::Version.new("5.1.0")
      RSpec.describe CallTemplate do
        describe "#make_lambda" do
          it "returns a Proc that calls the rails logger" do
            callback = double("mock callback")
            cb = described_class.build(:banana, callback)

            # Expect `make_lambda` to return a `Proc`.
            lambda_with_log = cb.make_lambda
            expect(lambda_with_log).to be_a(::Proc)

            # Mock the rails logger. We're going to expect it to receive a `debug` message.
            logger = double
            allow(logger).to receive(:debug)
            allow(::RailsCallbackLog).to receive(:logger).and_return(logger)
            target = double

            # Symbol-based callbacks have a `target` and a `value`. The definition of
            # these arguments is not clear to me yet. My guess is an AR callback's `target`
            # would be a model instance, for example.
            allow(target).to receive(:kiwi).with(:banana)
            allow(target).to receive(:banana)

            # Call our lambda and expect it to send a debug message to our mock logger.
            lambda_with_log.call(target, :kiwi)
            expect(logger).to have_received(:debug).once.with("Callback: :banana")
            expect(target).to have_received(:banana).once
          end
        end
      end
    else
      RSpec.describe Callback do
        describe "#make_lambda" do
          it "returns a Proc that calls the rails logger" do
            # Instantiate a `Callback`. For this test, none of the arguments matter.
            name = double.as_null_object
            filter = double.as_null_object
            kind = double.as_null_object
            options = double.as_null_object
            chain_config = double.as_null_object
            cb = described_class.new(name, filter, kind, options, chain_config)

            # Expect `make_lambda` to return a `Proc`.
            lambda_with_log = cb.make_lambda(:banana)
            expect(lambda_with_log).to be_a(::Proc)

            # Mock the rails logger. We're going to expect it to receive a `debug` message.
            logger = double
            allow(logger).to receive(:debug)
            allow(::Rails).to receive(:logger).and_return(logger)
            target = double

            # Symbol-based callbacks have a `target` and a `value`. The definition of
            # these arguments is not clear to me yet. My guess is an AR callback's `target`
            # would be a model instance, for example.
            allow(target).to receive(:kiwi).with(:banana)
            allow(target).to receive(:banana)

            # Call our lambda and expect it to send a debug message to our mock logger.
            lambda_with_log.call(target, :kiwi)
            expect(logger).to have_received(:debug).once.with("Callback: banana")
            expect(target).to have_received(:banana).once
          end
        end
      end
    end
  end
end


lib/rails_callback_log.rb
require "active_support/callbacks"

module RailsCallbackLog
  SOURCE_LOCATION_FILTERS = %w(app lib).map { |dir| (::Rails.root + dir).to_s }.freeze
  VERSION = "0.0.1"

  def self.matches_filter?(str)
    SOURCE_LOCATION_FILTERS.any? { |f| str.start_with?(f) }
  end
end

if ::Gem::Requirement.new("~> 4.2.0").satisfied_by?(::Rails.gem_version)
  module ActiveSupport
    module Callbacks
      class Callback
        def make_lambda_with_log(filter)
          original_lambda = make_lambda_without_log(filter)
          lambda { |*args|
            if caller.any? { |line| ::RailsCallbackLog.matches_filter?(line) }
              ::Rails.logger.debug(format("Woo Callback: %s", filter))
            end
            original_lambda.call(*args)
          }
        end
        alias_method_chain :make_lambda, :log
      end
    end
  end
else
  warn "ActiveSupportCallbackLog does not support rails version: #{::Rails.gem_version}"
end


lib/rails-callback_log.rb
require "rails_callback_log/version"

# Make sure `ActiveSupport::Callbacks` is loaded before we continue.
require "active_support/all"

module RailsCallbackLog
  # Filtering is very expensive. It makes my test suite more than 50%
  # slower. So, it's off by default.
  FILTER = ENV["RAILS_CALLBACK_LOG_FILTER"].present?.freeze

  class << self
    def logger
      ::Rails.logger || ::Logger.new(STDOUT)
    end

    def matches_filter?(str)
      source_location_filters.any? { |f| str.start_with?(f) }
    end

    def log(msg)
      if !FILTER || caller.any? { |line| matches_filter?(line) }
        logger.debug(format("Callback: %s", msg.inspect))
      end
    end

    private

    def source_location_filters
      @@filters ||= %w(app lib).map { |dir| (::Rails.root + dir).to_s }
    end
  end

  # In rails 5.1, we extend `CallTemplate`.
  module CallTemplateExtension
    # Rails 5.1 and above call `expand` to get a reference to the object and
    # method of the callback to execute. If the callback was a proc then
    # @override_block will be set. If the callback was an object then
    # @override_target will be set. If the callback was a symbol method name
    # then @method_name will be set.
    def expand(target, value, block)
      ::RailsCallbackLog.log(@override_block || @override_target || @method_name)

      super(target, value, block)
    end
  end

  # In rails 4.2 and 5.0, we extend `Callback`.
  module CallbackExtension
    # Returns a lambda that wraps `super`, adding logging.
    def make_lambda(filter)
      original_lambda = super(filter)
      lambda { |*args, &block|
        ::RailsCallbackLog.log(filter)

        original_lambda.call(*args, &block)
      }
    end
  end
end

# Install our `CallbackExtension` using module prepend.
module ActiveSupport
  module Callbacks
    if ::ActiveSupport.gem_version >= ::Gem::Version.new("5.1.0")
      class CallTemplate
        prepend ::RailsCallbackLog::CallTemplateExtension
      end
    else
      class Callback
        prepend ::RailsCallbackLog::CallbackExtension
      end
    end
  end
end
